executor/join.go

```diff
@@ -209,9 +282,124 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
 	}
 }
 
-// prepare runs the first time when 'Next' is called, it starts one worker goroutine to fetch rows from the big table,
-// and reads all data from the small table to build a hash table, then starts multiple join worker goroutines.
-func (e *HashJoinExec) prepare(goCtx goctx.Context) error {
+// fetchOuterChunks get chunks from fetches chunks from the big table in a background goroutine
+// and sends the chunks to multiple channels which will be read by multiple join workers.
+func (e *HashJoinExec) fetchOuterChunks(goCtx goctx.Context) {
+	defer func() {
+		for i := range e.outerResultChs {
+			close(e.outerResultChs[i])
+		}
+		e.workerWaitGroup.Done()
+	}()
+	for {
+		if e.finished.Load().(bool) {
+			return
+		}
+		var outerResource *outerChkResource
+		ok := true
+		select {
+		case <-e.closeCh:
```

Comment:
Can this case be removed?
When the executor is closed the outerChkResouceCh will be closed.